File: atlas_d_original_r7/source/verify.py
"""Browser/pose verification is not engineering validation."""
from pathlib import Path
import os,json,time
from playwright.sync_api import sync_playwright
BASE=os.getenv('BASE_URL','http://127.0.0.1:8787/index.html');OUT=Path(os.getenv('REVIEW_DIR','/tmp/atlas7/review'));OUT.mkdir(parents=True,exist_ok=True)
report={'url':BASE,'passed':False,'engineering_validated':False,'physical_iPhone_test':False,'tests':[]}
with sync_playwright() as pw:
 browser=pw.chromium.launch(headless=True,args=['--no-sandbox','--use-angle=swiftshader','--enable-unsafe-swiftshader'])
 for label,width,height in [('desktop',1600,1000),('phone',430,932)]:
  page=browser.new_page(viewport={'width':width,'height':height},device_scale_factor=1);errors=[];bad=[]
  page.on('pageerror',lambda e:errors.append(str(e)));page.on('requestfailed',lambda r:bad.append({'url':r.url,'error':r.failure}))
  res=page.goto(BASE,wait_until='domcontentloaded',timeout=60000);host=False
  if page.get_by_role('button',name='Open the page',exact=True).count():host=True;page.get_by_role('button',name='Open the page',exact=True).click()
  page.wait_for_function('window.ATLAS_R7?.ready === true',timeout=100000);page.wait_for_timeout(900);snapshots=[]
  def shot(name,p,view=None):
   page.evaluate('(p)=>{ATLAS_R7.pose(p)}',p)
   if view:page.evaluate('(v)=>ATLAS_R7.setView(v)',view)
   page.wait_for_timeout(240);page.screenshot(path=str(OUT/f'{label}_{name}.png'),timeout=25000);d=page.evaluate('ATLAS_R7.inspect()');snapshots.append(d);return d
  a=shot('surface',0,'hero');b=shot('ready',.8)
  assert b['depth']==0 and b['skyDrop']>6.2 and b['bridgeDrop']>4.6
  for k in ['tenderChase','tenderUtility']:assert -41<b[k][0]<-28 and abs(b[k][2])<4.5 and b[k][1]>1.25
  assert b['scaleTracks']==0 and b['mainHullVisible']
  shot('halfway',.53,'hero');shot('profile_surface',0,'profile');shot('profile_ready',.8);shot('recovery',.195,'recovery');shot('roofs',.565,'roofs');d=shot('submerged',1,'underwater')
  assert abs(d['depth']-26)<1e-5 and all(v<-24 for v in [d['tenderChase'][1],d['tenderUtility'][1]])
  assert d['mainHullScale']==[1,1,1] and d['pressureCoreScale']==[1,1,1]
  page.evaluate('ATLAS_R7.pose(.8)');page.locator('#core').click();page.wait_for_timeout(300);page.screenshot(path=str(OUT/f'{label}_bays.png'));page.locator('#core').click();page.locator('#reset').click();page.locator('#compare').click();assert abs(page.evaluate('ATLAS_R7.progress')-.8)<1e-5
  page.locator('#compare').click();assert page.evaluate('ATLAS_R7.progress')==0
  page.evaluate('ATLAS_R7.pose(.796)');page.locator('#play').click();page.wait_for_function('!ATLAS_R7.playing',timeout=15000);assert abs(page.evaluate('ATLAS_R7.progress')-.8)<1e-5
  page.locator('#timeline').fill('530');page.locator('#timeline').dispatch_event('input');assert abs(page.evaluate('ATLAS_R7.progress')-.53)<1e-5
  page.locator('#basis').click();assert page.locator('#notes').is_visible();page.locator('#close-notes').click();assert not errors,errors
  assetbad=[x for x in bad if any(n in x['url'] for n in ['ATLAS_D_Original.glb','runtime.js','viewer.js','style.css'])];assert not assetbad,assetbad
  report['tests'].append({'label':label,'viewport':[width,height],'http_status':res.status,'host_confirmation_required':host,'poses':snapshots,'errors':errors,'asset_failures':assetbad,'controls':['surface/ready fixed camera','conversion stop above water','timeline','recovery camera','shells camera','x-ray bays','design limits'],'passed':True});(OUT/'validation.json').write_text(json.dumps(report,indent=2));page.close()
 browser.close()
report['passed']=True;(OUT/'validation.json').write_text(json.dumps(report,indent=2));print(json.dumps({'passed':True,'url':BASE,'tests':len(report['tests'])}))

File: atlas_d_transform_r4/source/verify.mjs
import {chromium} from 'playwright';
import fs from 'node:fs';
const url=process.env.BASE_URL||'http://127.0.0.1:8764/index.html';
const dir=process.env.REVIEW_DIR||'/tmp/atlas-r4-review/local';fs.mkdirSync(dir,{recursive:true});
const result={url,revision:4,passed:false,physicalIPhoneTest:false,engineeringValidated:false,tests:[]};
const browser=await chromium.launch({headless:true,args:['--no-sandbox','--use-gl=angle','--use-angle=swiftshader','--enable-unsafe-swiftshader']});
try {
 for(const [label,width,height] of [['desktop',1440,1000],['phone',430,932]]) {
  const page=await browser.newPage({viewport:{width,height},deviceScaleFactor:1});
  const row={label,errors:[],assetFailures:[],poses:[]};result.tests.push(row);
  page.on('pageerror',e=>row.errors.push(String(e)));
  page.on('response',r=>{if(r.status()>=400&&/ATLAS_D_R4|runtime\.js|viewer\.js|style\.css/.test(r.url()))row.assetFailures.push({url:r.url(),status:r.status()});});
  await page.goto(url,{waitUntil:'domcontentloaded',timeout:90000});
  if(url.includes('githack')){
   await page.waitForTimeout(1500);
   const confirm=page.getByRole('button',{name:'Open the page',exact:true});
   if(await confirm.isVisible()){row.hostConfirmationRequired=true;await confirm.click();}
  }
  await page.waitForFunction(()=>window.ATLAS_D?.ready,{},{timeout:90000});
  for(const p of [0,.38,.62,1]){
   await page.evaluate(p=>{ATLAS_D.seek(p);ATLAS_D.setView('profile');},p);
   await page.waitForTimeout(700);
   row.poses.push(await page.evaluate(()=>{
    const a=ATLAS_D,roof=a.model.getObjectByName('Forward_leisure_canopy_roof'),t=a.model.getObjectByName('Starboard_terrace_hinge');
    return{progress:a.progress,depth:a.depth,roofLocalY:roof.position.y,terraceQuaternion:t.quaternion.toArray(),scaleTracks:a.clip.tracks.filter(t=>t.name.endsWith('.scale')).length,modelMeshes:a.renderer.info.render.calls,...a.diagnostics()};
   }));
   await page.screenshot({path:`${dir}/${label}_${p===0?'surface':p===.62?'ready':p===1?'submerged':'mid'}.png`,timeout:90000});
  }
  const drop=row.poses[0].roofLocalY-row.poses[2].roofLocalY;
  if(Math.abs(drop-5.16)>.01)throw new Error('Forward canopy did not physically lower 5.16 m');
  if(row.poses[2].depth>.01)throw new Error('Dive-ready comparison is not held at the surface');
  if(row.poses.some(p=>p.scaleTracks!==0))throw new Error('Structural scale tracks found');
  row.canopyDropMetres=drop;
  await page.evaluate(()=>ATLAS_D.reset());await page.locator('#compare').click();
  await page.waitForFunction(()=>Math.abs(ATLAS_D.progress-.62)<.002);
  await page.locator('#compare').click();await page.waitForFunction(()=>ATLAS_D.progress===0);
  // Test pause and the dedicated transformation stop without a long wall-clock wait.
  await page.locator('#transform').click();await page.waitForTimeout(1200);
  if(!await page.evaluate(()=>ATLAS_D.transitioning))throw new Error('Transform control did not start');
  await page.locator('#play').click();if(await page.evaluate(()=>ATLAS_D.transitioning))throw new Error('Pause failed');
  await page.locator('#transform').click();await page.evaluate(()=>{ATLAS_D.seek(.619);ATLAS_D.play()});await page.waitForFunction(()=>Math.abs(ATLAS_D.progress-.62)<.001&&!ATLAS_D.transitioning,{},{timeout:10000});row.automaticSurfaceStop=true;await page.evaluate(()=>{ATLAS_D.seek(.62);ATLAS_D.setView('mechanism')});
  await page.waitForTimeout(600);await page.screenshot({path:`${dir}/${label}_canopies.png`,timeout:90000});
  await page.locator('[data-view=stern]').click();await page.waitForTimeout(600);await page.screenshot({path:`${dir}/${label}_stern.png`,timeout:90000});
  await page.evaluate(()=>{ATLAS_D.setXray(true);ATLAS_D.setSystem('pressure')});await page.waitForTimeout(500);
  await page.screenshot({path:`${dir}/${label}_pressure.png`,timeout:90000});
  row.controls=['Before / Ready','Transform only','Pause','Canopies','Stern','Pressure-core inspection'];row.passed=row.errors.length===0&&row.assetFailures.length===0;
  if(!row.passed)throw new Error('Browser or model asset errors');
  await page.close();
 }
 result.passed=true;
} catch(e){result.failure=String(e);console.error(e);process.exitCode=1;}
finally{fs.writeFileSync(`${dir}/validation.json`,JSON.stringify(result,null,2));await browser.close();}
